designs/2022-unused-disable-directive-flexible-config/README.md

```diff
@@ -0,0 +1,247 @@
+2. Adding `reportUnusedDisableDirectives: true` to the `.eslintrc.js` config file (warns, and thus easily ignored)
+3. Enabling the third-party rule [eslint-comments/no-unused-disable](https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/no-unused-disable.html) (errors, but requires a third-party plugin)
+
+These existing solutions have various downsides:
+
+1. Lack of discoverability.
+2. Requires manual configuration to enable and thus not commonly used.
+3. Inability to treat unused disable directives as errors in shareable configs using the first-party options.
+4. Inconsistency in behavior between the CLI option and config file option of the same name which is unintuitive and causes confusion.
+
+### New Design
+
+We propose a new "explicit severity" design where the user can set the severity level of unused disable directive reporting using either the CLI option or the config file option. This design has the following benefits:
+
+- The user should always be able to achieve the desired effect (`off` / `warn` / `error`), regardless of where they set the option from (CLI option, config file option, or even from a shareable config)
+- No confusing inconsistency where one of the options triggers errors but the other triggers warnings
+
+We will also change the default behavior to error on unused disable directives so that all users (unless they opt-out) will benefit from the detection (and automatic removal with `--fix`) of unused disable directives.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+How the new "explicit severity" design works:
+
+- The config and CLI options are both changed from booleans to instead accept standard severity level string values: `off`, `warn`, `error` (or a the corresponding number for each level)
+- If both options are present, the CLI option takes precedence over (overrides) the config option
+- If only one of the options is present, the value of the present option is used
+- If neither option is present, the default value (`error`) is used
+- Note: There's only one underlying setting, and the config and CLI options are just two different ways of controlling it
+
+To reduce complexity and avoid confusion, we will not allow boolean values to be provided anymore, except during a possible intermediary [phase](#phases).
+
+In addition to allowing an explicit severity level to be passed to the CLI option `--report-unused-disable-directives` (e.g. `--report-unused-disable-directives warn`), we will also allow the option to be provided without a value (e.g. `--report-unused-disable-directives`) (a common CLI shorthand) which will default to `error` severity, the same as today.
```

Comment:
I'm not sure this is allowed with our current system. Please double-check.

Reply:
`eslint --report-unused-disable-directives` works, I checked.

Reply:
I think the question here is can we [configure `optionator`](https://github.com/eslint/eslint/blob/main/lib/options.js) to allow `--report-unused-disable-directive` with and without arguments.

Another question is whether we should do that because the commands would become ambiguous. For example, `eslint --report-unused-disable-directives warn foo` could mean "lint directory `foo` and report unused disable directives as warnings" or "lint directories  `warn` and `foo` and report unused disable directives as errors".

Reply:
I continue to find it bizarre that nonzero tools accept double-dashed argument with values without an explicit `=` sign - does eslint do that?

Reply:
I updated the RFC to clarify that these are the allowed values for the CLI option:

- `--report-unused-disable-directives` - same behavior as today (`error`) (common CLI shorthand to omit the value)
- `--report-unused-disable-directives=true` - same behavior as today (`error`)
- `--report-unused-disable-directives=false` - same behavior as today (`off`)
- `--report-unused-disable-directives=error` - new value
- `--report-unused-disable-directives=warn` - new value
- `--report-unused-disable-directives=off` - new value

Based on some initial testing, this should work. We can pass something like `type: "Boolean|String|Number"` and `enum: ["true", "false", "error", "warn", "off", "0", "1", "2"]` to optionator.

Reply:
I think the other options don't require an equals sign? If so, this may be confusing to have one option that requires an equal signs and a bunch of others that don't.

Reply:
`--report-unused-disable-directives-severity [severity]` sounds good to me.

If `--report-unused-disable-directives-severity` takes precedence and doesn't require `--report-unused-disable-directives` to have effect, then it might be less confusing to throw an error if both options are provided.

Do we want to keep `--report-unused-disable-directives` for convenience or just for backwards compatibility? If it's just for backwards compatibility, it might be better to mark it deprecated and remove it at some point.

Reply:
I was thinking about keeping `--report-unused-disable-directives` around as-is. Why?
- For convenience, since it's much shorter than `--report-unused-disable-directives-severity error` which is becoming quite tedious
- Erroring, as the shorthand does, is likely what most users want
- Avoid a breaking change

That said, if we preferred to have only one option, and were okay with a breaking change, then I would keep the shorter option name `--report-unused-disable-directives` and simply change it from accepting a boolean to the severity string.

--- 

There's the other question of whether, if we decide on two options, we should allow both options to be provided together. You're probably right that we should throw an error since it's redundant. The only reason I was thinking we could allow both is for this override situation:

```json
{
  "lint:js": "eslint --report-unused-disable-directives path/to/lint"
}
```

While playing around, the user could easily override this with `yarn lint:js --report-unused-disable-directives-severity warn` as needed. But I'm not attached to this edge case and I'm open to just throwing an exception to avoid any confusion.

Reply:
To take a step back, if the default is warn on unused disable directives, then what would `--report-unused-disable-directives` do? Because it only turns on this warning, it seems like it's actually not useful anymore? Or am I missing something?

It seems like we'd just want a flat to either up the severity or turn it off?

Reply:
Wouldn't the CLI flag be used to override a config that turned it off?

Reply:
`--report-unused-disable-directives` would continue to report errors as it does today. I have not proposed changing that default as I prefer this option to error.

To summarize what I'm proposing currently for CLI options and the use cases:
1. `--report-unused-disable-directives` - existing option to show errors (no change)
2. `--report-unused-disable-directives-severity error` redundant with above option but kept for completeness
3. `--report-unused-disable-directives-severity warn` for overriding `reportUnusedDisableDirectives: false` from a config file
4. `--report-unused-disable-directives-severity off` for overriding `reportUnusedDisableDirectives: true` from a config file (note that this RFC also makes warning the default behavior)

Reply:
Okay, I think it would help to clean up the language around this because I had to re-read what you wrote a few times before I got it.

The default with no flags is to report as warnings.

When using `--report-unused-disable-directives`, the severity is upped to error from warn.

When using `--report-unused-disable-directives-severity`, you can specifically override the severity.

Do I have that correct?

Reply:
Yes. I need to update the RFC to reflect this design, assuming we are happy with it.

Reply:
Makes sense to me.

Reply:
Makes sense to me too.

I also think we should throw an error if both options are provided.

Reply:
I have updated the RFC to describe the new design we agreed on here. Sorry for the delay.